src/validate/validate_schema.ts

```diff
@@ -0,0 +1,224 @@
+import {ValidationError} from '../error/validation_error';
+import {getType} from '../util/get_type';
+import {deepUnbundle, unbundle} from '../util/unbundle_jsonlint';
+import {isObjectLiteral} from '../util/is_object_literal';
+
+interface ValidateSchemaOptions {
+    key: string;
+    value: unknown;
+}
+
+export function validateSchema({
+    key,
+    value,
+}: ValidateSchemaOptions): ValidationError[] {
+    if (!isObjectLiteral(value)) {
+        return [
+            new ValidationError(
+                key,
+                value,
+                `object expected, ${getType(value)} found`
+            ),
+        ];
+    }
+    if (value.type === undefined && value.enum === undefined) {
+        return [
+            new ValidationError(
+                key,
+                value,
+                'schema must have a "type" or "enum" property'
+            ),
+        ];
+    }
+    if (value.default === undefined) {
+        return [new ValidationError(`${key}`, value, 'default is required')];
+    }
+
+    switch (unbundle(value.type)) {
+        case 'string':
+            return validateString(value, key);
+        case 'number':
+            return validateNumber(value, key);
+        case 'boolean':
+            return validateBoolean(value, key);
+        case 'array': {
+            return validateArray(value, key);
+        }
+        default: {
+            if (value.type !== undefined) {
+                return [
+                    new ValidationError(
+                        'type',
+                        value.type,
+                        'expected string, number or boolean'
+                    ),
+                ];
+            }
+        }
+    }
+
+    return validateEnum(value, key);
+}
+
+function validateString(schema: Record<string, unknown>, key: string) {
```

Comment:
Don't we have these kinds of validations already as part of the style spec validation code...?

Reply:
Yes and no :)

First, representing this structure in v8 schema was problematic, for example "type" keyword is restricted in v8 and in state it's simply a property name with a fixed value.
Secondly, my validators are referencing other properties, or are more complex and would require custom validators anyway - like enum, array, or number with minimum and maximum.

I tried to use built in validation system at first, but it seemed not fit for the job without major adjustments. 

I welcome suggestions from you and others more experienced with this code base.

Reply:
To be honest, I would like to avoid introducing JSON schema as part of this feature.
I would like to know what the limitations were of using the current type system?